crates/accelerate/src/commutation_analysis.rs

```diff
@@ -0,0 +1,197 @@
+
+use pyo3::types::{PyDict, PyList};
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+// custom types: IndexSet allows to iterate over the elements in insertion
+// order, but uses the std hasher, which is slower than ahash
+type AIndexSet<T> = IndexSet<T, ::ahash::RandomState>;
+type CommutingNodes = Vec<AIndexSet<NodeIndex>>;
+
+/// Compute the commutation sets for a given DAG.
+///
+/// We return two HashMaps:
+///  * {wire: commutation_sets}: For each wire, we keep a vector of index sets, where each index
+///     set contains mutually commuting nodes. Note that these include the input and output nodes
+///     which do not commute with anything.
+///  * {(node, wire): index}: For each (node, wire) pair we store the index indicating in which
+///     commutation set the node appears on a given wire.
+///
+/// For example, if we have a circuit
+///
+///     |0> -- X -- SX -- Z (out)
+///      0     2    3     4   1   <-- node indices including input (0) and output (1) nodes
+///
+/// Then we would have
+///
+///     commutation_set = {0: [[0], [2, 3], [4], [1]]}
+///     node_indices = {(0, 0): 0, (1, 0): 3, (2, 0): 1, (3, 0): 1, (4, 0): 2}
+///
+#[allow(clippy::type_complexity)]
+fn analyze_commutations_inner(
+    py: Python,
+    dag: &mut DAGCircuit,
+    commutation_checker: &mut CommutationChecker,
+) -> PyResult<(
+    HashMap<Wire, CommutingNodes>,
+    HashMap<(NodeIndex, Wire), usize>,
+)> {
+    let mut commutation_set: HashMap<Wire, CommutingNodes> = HashMap::new();
+    let mut node_indices: HashMap<(NodeIndex, Wire), usize> = HashMap::new();
+
+    let max_num_qubits = 3;
+
+    for qubit in 0..dag.num_qubits() {
+        let wire = Wire::Qubit(Qubit(qubit as u32));
+
+        for current_gate_idx in dag.nodes_on_wire(py, &wire, false) {
+            // get the commutation set associated with the current wire, or create a new
+            // index set containing the current gate
+            let commutation_entry = commutation_set
+                .entry(wire.clone())
+                .or_insert_with(|| vec![AIndexSet::from_iter([current_gate_idx])]);
+
+            // we can unwrap as we know the commutation entry has at least one element
+            let last = commutation_entry.last_mut().unwrap();
+
+            // if the current gate index is not in the set, check whether it commutes with
+            // the previous nodes -- if yes, add it to the commutation set
+            if !last.contains(&current_gate_idx) {
+                let mut all_commute = true;
+
+                for prev_gate_idx in last.iter() {
+                    // if the node is an input/output node, they do not commute, so we only
+                    // continue if the nodes are operation nodes
+                    if let (NodeType::Operation(packed_inst0), NodeType::Operation(packed_inst1)) =
+                        (&dag.dag[current_gate_idx], &dag.dag[*prev_gate_idx])
```

Comment:
Can `current_gate_idx` ever point to a non-op node? Not that it really matters because we have need to get the inner packed instruction, I was just wondering from a logic PoV.

Reply:
yep it even has to, since the commutation sets also contain indices for the input and output nodes (we iterate over `dag.nodes_on_wire(..., only_ops: false)`)